File: demo-android/app/src/androidTest/java/com/example/demo/e2e/interaction/validation/TaskValidator.kt
package com.example.demo.e2e.interaction.validation

import androidx.compose.ui.test.assertIsDisplayed
import androidx.compose.ui.test.assertTextEquals
import androidx.compose.ui.test.junit4.AndroidComposeTestRule
import com.example.demo.MainActivity
import com.example.demo.e2e.data.TaskData
import com.example.demo.e2e.interaction.page.InfoTaskModal
import io.qameta.allure.kotlin.Allure

class TaskValidator(
    rule: AndroidComposeTestRule<*, MainActivity>,
) {
    private val infoTaskModal = InfoTaskModal(rule)

    fun data(taskData: TaskData): TaskValidator {
        Allure.step("Validate task info data") {
            infoTaskModal.title().assertTextEquals(taskData.title)
            infoTaskModal.descriptionField().assertTextEquals(taskData.description)
            infoTaskModal.statusTag(taskData.status).assertIsDisplayed()
            infoTaskModal.priorityTag(taskData.priority).assertIsDisplayed()
        }
        return this
    }

    fun isValid() {
        Allure.step("Validate task is marked as valid") {
            infoTaskModal.validIcon().assertIsDisplayed()
        }
    }
}

File: demo-ios/DemoUITests/Interaction/Validation/TaskValidator.swift
import XCTest

final class TaskValidator {
    private let infoTaskModal: InfoTaskModal

    init(app: XCUIApplication) {
        infoTaskModal = InfoTaskModal(app: app)
    }

    @discardableResult
    func data(_ taskData: TaskData) -> TaskValidator {
        Allure.step("Validate task info data") {
            infoTaskModal.title().waitUntilExists()
            XCTAssertEqual(infoTaskModal.title().label, taskData.title)
            XCTAssertEqual(infoTaskModal.descriptionField().label, taskData.description)
            XCTAssertTrue(infoTaskModal.statusTag(taskData.status).exists)
            XCTAssertTrue(infoTaskModal.priorityTag(taskData.priority).exists)
        }
        return self
    }

    func isValid() {
        Allure.step("Validate task is marked as valid") {
            infoTaskModal.validIcon().waitUntilExists()
        }
    }
}
